### PR TITLE
Bh/cmp chunks

### PR DESCRIPTION
**Pull request by bheni**

### CODE REVIEW

File: go/store/nbs/cmp_chunk_table_writer.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+	"os"
+	"sort"
+
+	"github.com/golang/snappy"
+
+	"github.com/liquidata-inc/dolt/go/libraries/utils/iohelp"
+)
+
+// A ByteSink is an interface for writing bytes which can later be flushed to a writer
+type ByteSink interface {
```
Comment (zachmu) [inline go/store/nbs/cmp_chunk_table_writer.go]: Isn't this just a io.BufferedWriter?

File: go/store/nbs/cmp_chunk_table_writer.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+	destLen := len(dest)
+	srcLen := len(src)
+
+	if destLen < srcLen {
+		return 0, ErrBufferFull
+	}
+
+	n := copy(dest, src)
+
+	if n != srcLen {
```
Comment (zachmu) [inline go/store/nbs/cmp_chunk_table_writer.go]: Is this even possible after the above checks?

File: go/store/nbs/cmp_chunk_table_writer.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+	sink.pos += uint64(n)
+	return srcLen, nil
+}
+
+// Flush writes all the data that was written to the ByteSink to the supplied writer
+func (sink *FixedBufferByteSink) Flush(wr io.Writer) error {
+	return iohelp.WriteAll(wr, sink.buff[:sink.pos])
+}
+
+// BlockBufferByteSink allocates blocks of data which of a given block size to store the bytes written to the sink. New
```
Comment (zachmu) [inline go/store/nbs/cmp_chunk_table_writer.go]: typo

File: go/store/nbs/cmp_chunk_table_writer.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+const defaultTableSinkBlockSize = 2 * 1024 * 1024
+
+// ErrNotFinished is an error returned by a CmpChunkTableWriter when a call to Flush* is called before Finish is called
+var ErrNotFinished = errors.New("Not finished")
+
+// ErrAlreadyFinished is an error returned if Finish is called more than once on a CmpChunkTableWriter
+var ErrAlreadyFinished = errors.New("Already Finished")
+
+// CmpChunkTableWriter writes CompressedChunks to a table file
+type CmpChunkTableWriter struct {
```
Comment (zachmu) [inline go/store/nbs/cmp_chunk_table_writer.go]: Much of the IO code below here would benefit from compressing the error handling to inlined if statements where possible.

File: go/store/nbs/cmp_chunk_table_writer_test.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	for i := 0; i < 32; i++ {
+		_, err := suite.sink.Write(data)
+		assert.NoError(suite.t, err)
+	}
+}
+
+func TestCmpChunkTableWriter(t *testing.T) {
+	// Put some chunks in a table file and get the buffer back which cotains the table file data
```
Comment (zachmu) [inline go/store/nbs/cmp_chunk_table_writer_test.go]: typo

File: go/store/nbs/cmp_chunk_table_writer_test.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+	compareContentsOfTables(t, ctx, hashes, tr, outputTR)
+}
+
+func compareContentsOfTables(t *testing.T, ctx context.Context, hashes hash.HashSet, expectedRd, actualRd tableReader) {
+	expected, err := readAllChunks(ctx, hashes, expectedRd)
+	require.NoError(t, err)
+	actual, err := readAllChunks(ctx, hashes, actualRd)
+	require.NoError(t, err)
+
+	assert.Equal(t, len(expected), len(actual))
```
Comment (zachmu) [inline go/store/nbs/cmp_chunk_table_writer_test.go]: You don't need this check, the following one handles length as well as content

File: go/store/nbs/table_reader.go [hunk 7]
```diff
@@ -395,7 +494,31 @@ func (tr tableReader) getManyAtOffsets(
...
+}
 
+func (tr tableReader) getManyAtOffsetsWithReadFunc(
+	ctx context.Context,
+	reqs []getRecord,
+	offsetRecords offsetRecSlice,
+	wg *sync.WaitGroup,
+	ae *atomicerr.AtomicError,
+	stats *Stats,
+	readAtOffsets func(
```
Comment (zachmu) [inline go/store/nbs/table_reader.go]: Might want to declare a type for this to save the verbosity here

File: go/store/nbs/table_reader.go [hunk 8]
```diff
@@ -596,13 +700,20 @@ func (tr tableReader) extract(ctx context.Context, chunks chan<- extractRecord)
 
 	sendChunk := func(i uint32) error {
 		localOffset := tr.offsets[i] - tr.offsets[0]
-		data, err := tr.parseChunk(buff[localOffset : localOffset+uint64(tr.lengths[i])])
+
+		cmp, err := NewCompressedChunk(hash.Hash(hashes[i]), buff[localOffset:localOffset+uint64(tr.lengths[i])])
```
Comment (zachmu) [inline go/store/nbs/table_reader.go]: What's the purpose of the change to this function? Seems like equivalent to the old version, but not sure.

File: go/store/nbs/cmp_chunk_table_writer.go (hunk 2 above)
Comment (bheni) [reply go/store/nbs/cmp_chunk_table_writer.go]: removed

File: go/store/nbs/cmp_chunk_table_writer.go (hunk 3 above)
Comment (bheni) [reply go/store/nbs/cmp_chunk_table_writer.go]: fixed

File: go/store/nbs/cmp_chunk_table_writer.go [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+	if err != nil {
+		return "", err
+	}
+
+	var h []byte
+	h = blockHash.Sum(h)
+
+	var blockAddr addr
+	copy(blockAddr[:], h)
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/store/nbs/cmp_chunk_table_writer.go]: because copy requires a slice, not an array.  This takes the entire array as a slice

File: go/store/nbs/cmp_chunk_table_writer.go [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+		return ErrNotFinished
+	}
+
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.ModePerm)
+
+	if err != nil {
+		return err
+	}
+
+	err = tw.sink.Flush(f)
```
[the remark this answers is not in the document]
Comment (bheni) [reply go/store/nbs/cmp_chunk_table_writer.go]: fixed

File: go/store/nbs/cmp_chunk_table_writer_test.go (hunk 5 above)
Comment (bheni) [reply go/store/nbs/cmp_chunk_table_writer_test.go]: fixed

File: go/store/nbs/cmp_chunk_table_writer.go (hunk 4 above)
Comment (bheni) [reply go/store/nbs/cmp_chunk_table_writer.go]: I don't like that style

File: go/store/nbs/table_reader.go (hunk 8 above)
Comment (bheni) [reply go/store/nbs/table_reader.go]: parseChunk has been removed, and everything uses the CompressedChunk class now so that I don't have to maintain 2 code paths that do the same thing.

File: go/store/nbs/cmp_chunk_table_writer_test.go (hunk 6 above)
Comment (bheni) [reply go/store/nbs/cmp_chunk_table_writer_test.go]: You are right, I don't need it, but I think it's faster to see what went wrong, and provides disambiguation of what went wrong.
